src/main/scala/ir/cilvisitor/CILVisitor.scala

```diff
@@ -0,0 +1,196 @@
+    def continue(v: CILVisitorImpl, l: MemoryLoad): MemoryLoad = {
+      MemoryLoad(visit_mem(l.mem), visit_expr(l.index), l.endian, l.size)
+    }
+    doVisit(v, v.vload(n), n, continue)
+  }
+
+  def visit_var(n: Variable): Variable = {
+    doVisit(v, v.vvar(n), n, (_, n) => n)
+  }
+
+  def visit_lvar(n: Variable): Variable = {
+    doVisit(v, v.vlvar(n), n, (_, n) => n)
+  }
+
+  def visit_mem(n: Memory): Memory = {
+    doVisit(v, v.vmem(n), n, (_, n) => n)
+  }
+
+  def visit_lmem(n: Memory): Memory = {
+    doVisit(v, v.vlmem(n), n, (_, n) => n)
+  }
+
+  def visit_jump(j: Jump): Jump = {
+    doVisit(v, v.vjump(j), j, (_, j) => j)
+  }
+
+  def visit_fallthrough(j: Option[GoTo]): Option[GoTo] = {
+    doVisit(v, v.vfallthrough(j), j, (_, j) => j)
+  }
+
+  def visit_expr(n: Expr): Expr = {
+    def continue(v: CILVisitorImpl, n: Expr): Expr = n match {
+      case n: Literal                           => n
+      case MemoryLoad(mem, index, endian, size) => MemoryLoad(visit_mem(mem), visit_expr(index), endian, size)
+      case Extract(end, start, arg)             => Extract(end, start, visit_expr(arg))
+      case Repeat(repeats, arg)                 => Repeat(repeats, visit_expr(arg))
+      case ZeroExtend(bits, arg)                => ZeroExtend(bits, visit_expr(arg))
+      case SignExtend(bits, arg)                => SignExtend(bits, visit_expr(arg))
+      case BinaryExpr(op, arg, arg2)            => BinaryExpr(op, visit_expr(arg), visit_expr(arg2))
+      case UnaryExpr(op, arg)                   => UnaryExpr(op, visit_expr(arg))
+      case v: Variable                          => visit_var(v)
+      case UninterpretedFunction(n, params, rt) => UninterpretedFunction(n, params.map(visit_expr), rt)
+    }
+    doVisit(v, v.vexpr(n), n, continue)
+  }
+
+  def visit_stmt(s: Statement): List[Statement] = {
+    def continue(v: CILVisitorImpl, n: Statement) = n match {
+      case m: MemoryAssign => {
+        m.mem = visit_lmem(m.mem)
+        m.index = visit_expr(m.index)
+        m.value = visit_expr(m.value)
+        m
+      }
+      case m: Assign => {
+        m.rhs = visit_expr(m.rhs)
+        m.lhs = visit_lvar(m.lhs)
+        m
+      }
+      case s: Assert => {
+        s.body = visit_expr(s.body)
+        s
+      }
+      case s: Assume => {
+        s.body = visit_expr(s.body)
+        s
+      }
+      case n: NOP => n
+    }
+    doVisitList(v, v.vstmt(s), s, continue)
+  }
+
+  def visit_block(b: Block): Block = {
+    def continue(v: CILVisitorImpl, b: Block) = {
+      b.statements.foreach(s => {
+        val r = visit_stmt(s)
+        r match {
+          case Nil      => b.statements.remove(s)
+          case n :: Nil => b.statements.replace(s, n)
+          case n :: tl => {
+            b.statements.replace(s, n)
+            var p = n
+            for (i <- tl) {
+              b.statements.insertAfter(p, i)
+              p = i
+            }
+          }
+        }
+      })
```

Comment:
a bit ceremonious. i assume this is required by intrusive list.

would this be useful to move into a general function for intrusive lists?